Add configuration `SSHAccess` for workers settings (#7188)
gardener/gardener: docs/usage/shoot_workers_settings.md

> # Configure all shoot workers

Reviewer: ```suggestion
# Shoot Worker Nodes Settings
```

Also add this doc to the index (`README.md`).

> Users can configure settings, that affect all worker nodes, in the `Shoot` API via `.spec.provider.workersSettings`.

Reviewer: ```suggestion
Users can configure settings affecting all worker nodes via `.spec.provider.workersSettings` in the `Shoot` resource.
```

> 
> ## EnsureSSHAccessDisabled

Reviewer: ```suggestion

## SSH Access
```

> `EnsureSSHAccessDisabled` indicates whether the `sshd.service` on the worker nodes should be disabled. When set to `true` a service called `sshddisabler.service` is started on each worker which runs a script every 15 seconds in order to ensure that the `sshd.service` is stopped and disabled. It also terminates all established ssh connections.

Reviewer: ```suggestion
`EnsureSSHAccessDisabled` indicates whether the `sshd.service` on the worker nodes should be disabled. When set to `true`, a systemd service called `sshd-disabler.service` is started on each worker node. It runs a script every 15 seconds in order to ensure that the `sshd.service` is stopped and disabled. It also terminates all established SSH connections.
```

> `EnsureSSHAccessDisabled` is set to `false` by default and when it is disabled it does not guarantee that `sshd.service` is enabled or running. If this configuration is set to `true` then edited to `false` the `sshd.service` will not be enabled automatically even if it is enabled by default for the specific operating system.

Reviewer: > If this configuration is set to `true` then edited to `false` the `sshd.service` will not be enabled automatically even if it is enabled by default for the specific operating system.

Well, that's undesired, isn't it? My expectation as user is that I can freely enable SSH access and disable it again (and go back and forth how often I want), and I expect all nodes to pick up this change accordingly.

Reviewer 2: I think that this is debatable. Right now we do nothing if `EnsureSSHAccessDisabled=false` since the wording does not imply that we should enable ssh but only ensure that it is disabled.

My understanding is that it is not guaranteed that the `sshd.service` is installed and enabled on the underlying OS.
- If the ssh server is not installed then I do not think that `sshAccess.enabled=true` makes sense since we cannot enable something that is not present.
- If it is installed but the service is disabled for some reason then again I do not think that Gardener wants to tamper with that.

WDYT?

Reviewer: If we now start to make SSH access configurable then we should do it properly. To me, the current proposal sounds like a workaround for the non-established expectations w.r.t. SSH.

Implicitly, we always assumed that the OS image has an `sshd` service that is installed and running out of the box. While this is not strictly needed, we should probably still extend https://github.com/gardener/gardener/blob/master/docs/extensions/operatingsystemconfig.md with information about SSH.

With this documentation and the proposed `sshAccess.enabled` field, it is much more intuitive for users/operators. If an OS image really does not have any SSH service (which is fine/totally valid), then it can serve a validating webhook (like other extensions like providers) to default this to `sshAccess.enabled=false` and to prevent users from changing it.

Reviewer 2: > If we now start to make SSH access configurable

This PR does not address this in its current state, but rather only handles disabling `ssh`. I understand your proposal is that we implement on/off handling for the `ssh` service. I am fine with this proposal but I also think that we should cover the case in which Gardener does not interfere with `ssh` at all. IMO three options should exist.

1. `sshAccess.enabled` is not set. This should be the default behavior and when this is the case Gardener should not tamper with any `ssh` settings on the nodes. This is also the current behavior. Users can already have custom `ssh` logic implemented and applied to nodes.
2. `sshAccess.enabled=true`. If a user sets this explicitly then Gardener enables the service and a validating admission webhook should ensure that the OS has an `ssh` service.
3. `sshAccess.enabled=false`. If a user sets this explicitly then Gardener disables the service and a validating admission webhook should ensure that the OS has an `ssh` service.

Reviewer: > sshAccess.enabled is not set. This should be the default behavior and when this is the case Gardener should not tamper with any ssh settings on the nodes. This is also the current behavior.

The current default is rather `sshAccess.enabled=true` (at least this is what we assume everywhere).

> Users can already have custom ssh logic implemented and applied to nodes.

WDYM by this?

Reviewer 2: > The current default is rather sshAccess.enabled=true (at least this is what we assume everywhere).

I do not think that we can 100% assume this since right now Gardener does not enable `ssh` explicitly. We can expect this to be the case, but nothing stops a user from deploying a daemonset that disables the `ssh` service on the nodes. Then if Gardener starts deploying a service that explicitly tries to enable `ssh`, it is a breaking change since we are trying to overwrite user defined behavior.

Reviewer: Well, with this argument we cannot ever extend the nodes or kubelet APIs anymore since users might have already interfered or changed configuration. :-) I still think that we should default it to `enabled=true` and notify users via release note that they should make use of this field and drop custom `DaemonSet`s (they would probably be happy about this new API since they wouldn't need to maintain this anymore themselves).
Sure, if they don't follow up timely, then SSH will be enabled again (if it was disabled before), but it's not too critical given that the nodes are not exposed to the public (and even if, you still need the right key).

Reviewer 2: > Well, with this argument we cannot ever extend the nodes or kubelet APIs anymore since users might have already interfered or changed configuration.

Yes, you are right and I am not suggesting that, but rather suggesting the more flexible (my opinion) option for the specific case.
1. Preserve the current Gardener behavior. If the user wants to handle the problem with a custom solution - so be it. There might be a use case where ssh access is wanted for a particular set of nodes and other nodes run with ssh disabled. Gardener will still allow that.
2. Allow the users to handle control over the ssh configs to Gardener. Then depending on the `true,false` value the respective config will be ensured.

We can still write the release note and inform users about this new API and I am sure that they will be happy about it. Then we give them full control and the ability to decide what they want to do with it. I do not get why we would want to remove the first option. Maybe I am missing something? @vlerenc maybe you can give a second (or rather third) opinion on the matter since we have discussed the topic in the past. Of course, feedback from other team members/contributors is also appreciated.

Don't get me wrong - the "strictly true/false" implementation is something that I can agree on if it is backed by the common opinion.

Reviewer: My motivation is reducing the complexity of the subject matter. Strict true/false is simple and easy to understand.
To me, this example where SSH access is only wanted for particular set of nodes is an artificial use case. Why would I do this/what would be the benefits?

As an user, either I want SSH to be completely disabled because I don't need it and want to be compliant with whatever standards, or I need it to be enabled (probably for short-term debugging purposes only). Given that nodes are private anyways and the key is still needed, there seems no value in managing an own `DaemonSet` for selecting certain nodes only.

TL;DR: Let's design this so that it's simple and easy-to-use unless we indeed find a realistic scenario with reasonable arguments justifying such "special behaviour".

Reviewer 3: I have not much context, so from what I read above, a few questions/notes:
- You call `sshAccess.enabled` default, because it's implicitly part of the OS configuration. Rafael says, it's rather enabled by default, because that's how our OSes are configured. I guess you are both right then.
- You suggest to have the first option why exactly? `sshd` is then enabled or not based on the OS configuration, that is not in the hands of the end users. That is not particularly useful. So, you want it so that an end user can tell Gardener to be hands-off and don't interfere with whatever logic the customer wants to apply to the nodes (possibly via a daemonset)?
- Instead of on-by-default and off-when-opted-out, you suggest three options: `hands-off`, `on` and `off`. Given our OS images/configurations, `hands-off` is then practically `on`. When you switch it `on` or `off`, you plan to invoke the appropriate `systemctl` commands, so that you are OS-independent?

Reviewer 2: I agree that strict true/false is simpler to implement and of course it is easier to understand since it implies two options instead of three.

> So, you want it so that an end user can tell Gardener to be hands-off and don't interfere with whatever logic the customer wants to apply to the nodes (possibly via a daemonset)?

Yes. I do not have a concrete use case for this, but rather I am suggesting to leave the possibility open.

>  Given our OS images/configurations, hands-off is then practically on.

Yes, unless some custom logic is applied by users (via daemonset or something else).

> When you switch it on or off, you plan to invoke the appropriate systemctl commands, so that you are OS-independent?

Yes.
- `hands-off` means a systemd service that does nothing and does not restart.
- `on` means a systemd service that continuously checks and ensures that the `sshd.service` is enabled and active.
- `off` means a systemd service that continuously checks and ensures that the `sshd.service` is disabled and not active. It will also terminate any established `ssh` connections.

Reviewer 2: Since I do not want to slow down the development on this feature I suggest that we proceed with what @rfranzke proposed, i.e., the strict `on/off` configuration. @AleksandarSavchev do you also agree with this proposition?

Author: Yes, I will begin implementing the `on/off` configuration.

# Shoot Worker Nodes Settings

Users can configure settings affecting all worker nodes via `.spec.provider.workersSettings` in the `Shoot` resource.

## SSH Access

> `SSHAccess` indicates whether the `sshd.service` on the worker nodes should be be enabled and running or disabled and stopped . When set to `true`, a systemd service called `sshd-ensurer.service` is started on each worker node. It runs a script every 15 seconds in order to ensure that the `sshd.service` is enabled and running. If it is set to `false`, the `sshd-ensurer.service` service ensures that `sshd.service` is stopped and disabled. It also terminates all established SSH connections, existing `Bastion` resources are deleted in shoot reconcilation and new ones are prevented from being created. SSH keypairs stopped being rotated, secrets with suffixes "ssh-keypair" and "ssh-keypair.old" are deleted and the gardener-user.service is not deployed to worker nodes.

Reviewer 2: ```suggestion
`SSHAccess` indicates whether the `sshd.service` should be running on the worker nodes. This is ensured by a systemd service called `sshd-ensurer.service` which runs on every 15 seconds on each worker node. When set to `true`, the systemd service ensures that the `sshd.service` is enabled and running. If it is set to `false`, the systemd service ensures that `sshd.service` is stopped and disabled. It also terminates all established SSH connections. In addition when this value is set to `false` existing `Bastion` resources are deleted during shoot reconciliation and new ones are prevented from being created, ssh keypairs are not created/rotated, ssh keypair secrets are deleted from the Garden cluster and the `gardener-user.service` is not deployed to the worker nodes.
```

> `SSHAccess` indicates whether the `sshd.service` should be running on the worker nodes. This is ensured by a systemd service called `sshd-ensurer.service` which runs on every 15 seconds on each worker node. When set to `true`, the systemd service ensures that the `sshd.service` is enabled and running. If it is set to `false`, the systemd service ensures that `sshd.service` is stopped and disabled. It also terminates all established SSH connections. In addition when this value is set to `false` existing `Bastion` resources are deleted during shoot reconciliation and new ones are prevented from being created, ssh keypairs are not created/rotated, ssh keypair secrets are deleted from the Garden cluster and the `gardener-user.service` is not deployed to the worker nodes.

Reviewer: ```suggestion
`SSHAccess` indicates whether the `sshd.service` should be running on the worker nodes. This is ensured by a systemd service called `sshd-ensurer.service` which runs every 15 seconds on each worker node. When set to `true`, the systemd service ensures that the `sshd.service` is enabled and running. If it is set to `false`, the systemd service ensures that `sshd.service` is stopped and disabled. This also terminates all established SSH connections. In addition, when this value is set to `false`, existing `Bastion` resources are deleted during `Shoot` reconciliation and new ones are prevented from being created, SSH keypairs are not created/rotated, SSH keypair secrets are deleted from the Garden cluster, and the `gardener-user.service` is not deployed to the worker nodes.
```

> `SSHAccess` is set to `true` by default.

Reviewer: ```suggestion
`sshAccess.enabled` is set to `true` by default.
```

# Configure all shoot workers

Users can configure settings, that affect all worker nodes, in the `Shoot` API via `.spec.provider.workersSettings`.

> Currently, we have the following fields in `WorkersSettings`:

Reviewer 2: ```suggestion
```

## EnsureSSHAccessDisabled

> `EnsureSSHAccessDisabled` indicates whether the worker nodes `sshd.service` is disabled. When set to `true` a service `sshddisabler.service` is started in each worker which runs a script every 15 seconds to stop and disable `sshd.service`, it also tereminates all already established ssh connections.

Reviewer 2: ```suggestion
`EnsureSSHAccessDisabled` indicates whether the `sshd.service` on the worker nodes should be disabled. When set to `true` a service called `sshddisabler.service` is started on each worker which runs a script every 15 seconds in order to ensure that the `sshd.service` is stopped and disabled. It also terminates all established ssh connections.
```

> `EnsureSSHAccessDisabled` is set to `false` by default and when it is disabled it does not guarantee that `sshd.service` is enabled. If this configuration is set to `true` then edited to `false` the `sshd.service` will not be enabled automatically even if it is enabled by default for the specific `os`.

Reviewer 2: ```suggestion
`EnsureSSHAccessDisabled` is set to `false` by default and when it is disabled it does not guarantee that `sshd.service` is enabled or running. If this configuration is set to `true` then edited to `false` the `sshd.service` will not be enabled automatically even if it is enabled by default for the specific operating system.
```
